core/src/main/java/com/google/zxing/qrcode/encoder/MinimalEncoder.java

```diff
@@ -0,0 +1,736 @@
+//- position-in-input: length of input in characters
+//- character-encoding: minimally 3 (ISO-8859-1, UTF-8 and UTF-16). The input is scanned up front and for example ISO-8859-6 is added if an arabic character is encountered.
+//- version-class: exactly 3 representing version 1-9, 10-26 and 27-40.
+//- Mode: exactly 4 (representing the modes KANJI, ALPHANUMERIC, NUMERIC and BYTE)
+    ResultNode[][][][] memoizedResults; //[position-in-input][character-encoding][version-class][Mode]
+
+/**Encoding is optional (default ISO-8859-1) and version is optional (minimal version is computed if not specified*/
+    MinimalEncoder(String stringToEncode,Version version,boolean isGS1) {
+        this.stringToEncode = stringToEncode;
+        if (version != null) {
+            this.version = version;
+        }
+        this.isGS1 = isGS1;
+        CharsetEncoder[] isoEncoders = new CharsetEncoder[15];
+        isoEncoders[0] = StandardCharsets.ISO_8859_1.newEncoder(); //shouldn't it be Shift_JIS?
+        //isoEncoders[0] = Encoder.DEFAULT_BYTE_MODE_ENCODING.newEncoder(); 
+        for (int i = 0; i < stringToEncode.length(); i++) {
+            int cnt = 0;
+            int j;
+            for (j = 0; j < 15; j++) {
+                if (isoEncoders[j] != null) {
+                    cnt++;
+                    if (isoEncoders[j].canEncode(stringToEncode.charAt(i))) {
+                        break;
+                    }
+                }
+            }
+            if (cnt == 14) { //we need all. Can stop looking further.
+                break;
+            }
+            if (j >= 15) { //no encoder found 
+                for (j = 0; j < 15; j++) {
+                    if (j != 11 && isoEncoders[j] == null) { // ISO-8859-12 doesn't exist
+                        try {
+                            CharsetEncoder ce = Charset.forName("ISO-8859-" + (j + 1)).newEncoder();
+                            if (ce.canEncode(stringToEncode.charAt(i))) {
+                                isoEncoders[j] = ce;
+                                break;
+                            }
+                        } catch (UnsupportedCharsetException e) { }
+                    }
+                }
+            }
+        }
+        int numberOfEncoders = 0;
+        for (int j = 0; j < 15; j++) {
+            if (isoEncoders[j] != null) {
+                numberOfEncoders++;
+            }
+        }
+        encoders = new CharsetEncoder[numberOfEncoders + 2];
+        int index = 0;
+        for (int j = 0; j < 15; j++) {
+            if (isoEncoders[j] != null) {
+                encoders[index++] = isoEncoders[j];
+            }
+        }
+        encoders[index++] = StandardCharsets.UTF_8.newEncoder();
+        encoders[index++] = StandardCharsets.UTF_16BE.newEncoder();
+        
+        memoizedResults = new ResultNode[stringToEncode.length()][encoders.length][3][4];
+    }
+    public static void main(String[] args) throws Exception {
```

Comment:
Space between methods
But we don't need a main() runner method here. The CLI is in another class